File: src/ios/Sources/AnsightToolsReflection/InvokeMethodTool.swift
import AnsightCore
import Foundation

public final class InvokeMethodTool: AnsightTool {
    private let runtime: AnsightRuntime
    private let options: AnsightReflectionToolsOptions

    public init(
        options: AnsightReflectionToolsOptions = .default,
        runtime: AnsightRuntime = .shared
    ) {
        self.runtime = runtime
        self.options = options
    }

    public var descriptor: AnsightToolDescriptor {
        AnsightToolDescriptor(
            id: AnsightReflectionToolIds.invokeMethod,
            name: "Invoke Method",
            description: "Invokes an opt-in method reachable from a registered iOS object root.",
            category: "reflect",
            policy: .critical,
            keywords: "reflection invoke method runtime",
            argumentsSchema: AnsightReflectionToolSchemas.invokeMethodArguments,
            resultSchema: AnsightReflectionToolSchemas.invokeMethodResult
        )
    }

    public func execute(arguments: [String: String]) throws -> AnsightToolExecutionResult {
        do {
            return .success(try AnsightReflectionSupport.invokeMethod(runtime: runtime, options: options, arguments: arguments))
        } catch {
            return .failure(error.localizedDescription, errorCode: "reflect_invoke_failed")
        }
    }
}

File: src/dotnet/Ansight.Tools.Reflection/InvokeMethodTool.cs
namespace Ansight.Tools.Reflection;

public sealed class InvokeMethodTool : ITool
{
    private readonly ReflectionToolsOptions options;

    public InvokeMethodTool(ReflectionToolsOptions? options = null)
    {
        this.options = options ?? ReflectionToolsOptions.Default;
    }

    public string Category => "reflect";

    public ToolPolicy Policy => ToolPolicy.Critical;

    public string Id => ReflectionToolIds.InvokeMethod;

    public string Name => "Invoke Method";

    public string Description => "Invokes an instance method reachable from a registered live object root.";

    public string Keywords => "reflection invoke method runtime";

    public ToolSchema ArgumentsSchema => ReflectionToolSchemas.InvokeMethodArguments;

    public ToolSchema ResultSchema => ReflectionToolSchemas.InvokeMethodResult;

    public Task<ToolResult> Execute(IReadOnlyDictionary<string, string> arguments)
    {
        ArgumentNullException.ThrowIfNull(arguments);

        try
        {
            return Task.FromResult(ToolResult.Success(ReflectionSupport.InvokeMethod(options, arguments)));
        }
        catch (Exception exception)
        {
            return Task.FromResult(ToolResult.Failure(exception.Message, errorCode: "reflect_invoke_failed"));
        }
    }
}
